File: src/spoj/dynamic_programming/WACHOVIA.c
// http://www.spoj.com/problems/WACHOVIA/

#include <stdio.h>
#include <memory.h>

#define MAX_WEIGHT 10001
#define MAX_BAGS_COUNT 51

#define NEGATIVE_INFINITY -10000000

typedef
struct {
    long weight;
    long cost;
} Bag;

Bag BAGS[MAX_BAGS_COUNT];

long MEMOIZED[MAX_BAGS_COUNT][MAX_WEIGHT];
int IS_MEMOIZED[MAX_BAGS_COUNT][MAX_WEIGHT];

inline long max(long a, long b) {
    return (a > b) ? a : b;
}

long calculateMaxRevenue(int currBagIndex, int maxWeight) {
    if(maxWeight < 0) {
        return NEGATIVE_INFINITY;
    }
    
    if(currBagIndex < 0) {
        return 0;
    }
    
    if(IS_MEMOIZED[currBagIndex][maxWeight]) {
        return MEMOIZED[currBagIndex][maxWeight];
    }
    
    MEMOIZED[currBagIndex][maxWeight] = max(calculateMaxRevenue(currBagIndex - 1, maxWeight),
                                            calculateMaxRevenue(currBagIndex - 1, maxWeight - BAGS[currBagIndex].weight) + BAGS[currBagIndex].cost);
    
    IS_MEMOIZED[currBagIndex][maxWeight] = 1;
    
    return MEMOIZED[currBagIndex][maxWeight];
}

int main() {
    int testCasesNum;
    int maxWeight;
    int bagsCount;
    int i;
    
    scanf("%d", &testCasesNum);
    while(testCasesNum--) {
        scanf("%d %d", &maxWeight, &bagsCount);
        for(i = 0; i < bagsCount; ++i) {
            scanf("%ld %ld", &BAGS[i].weight, &BAGS[i].cost);
        }
        memset(IS_MEMOIZED, 0, sizeof IS_MEMOIZED);
        
        printf("Hey stupid robber, you can get %ld.\n", calculateMaxRevenue(bagsCount - 1, maxWeight));
    }
    
    return 0;
}

File: src/spoj/dynamic_programming/WACHOVIA.java
package spoj.dynamic_programming;

// http://www.spoj.com/problems/WACHOVIA/

public class WACHOVIA {

	public static void main(String[] args) {
		System.out.println(calculateMaxCost(
				new Bag[]{
						new Bag(178, 12),
						new Bag(30, 1),
						new Bag(13, 7),
						new Bag(34, 8),
						new Bag(87, 6),
				}, 34));

		System.out.println(calculateMaxCost(
				new Bag[]{new Bag(900, 25)}, 900));

		System.out.println(calculateMaxCost(
				new Bag[]{
						new Bag(27, 16),
						new Bag(131, 9),
						new Bag(132, 17),
						new Bag(6, 5),
						new Bag(6, 23),
						new Bag(56, 21),
						new Bag(100, 25),
						new Bag(1, 25),
						new Bag(25, 25),
						new Bag(100, 2),
				}, 100));
	}
	static long calculateMaxCost(Bag[] bags, long maxWeight) {
		return calculateMaxCost(bags, bags.length - 1, maxWeight);
	}

	static final int NEGATIVE_INFINITY = -100000000;

	// Solution with exponential complexity
	// TODO: add memoization
	static long calculateMaxCost(Bag[] bags, int currBagIndex, long maxWeight) {

		if (maxWeight < 0) {
			return NEGATIVE_INFINITY;
		}

		if (currBagIndex < 0) {
			return 0;
		}

		Bag currBag = bags[currBagIndex];

		// memoize cost[bagIndex][maxWeight]
		return Math.max(
				calculateMaxCost(bags, currBagIndex - 1, maxWeight),
				calculateMaxCost(bags, currBagIndex - 1, maxWeight - currBag.weight) + currBag.cost);
	}

	static class Bag {
		long weight;
		long cost;
		Bag(long weight, long cost) {
			this.weight = weight;
			this.cost = cost;
		}
	}
}
